urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/utils/Configuration.java

```diff
@@ -0,0 +1,343 @@
+    {
+        return (List<String>) get(name);
+    }
+    
+    @SuppressWarnings("unchecked")
+    public List<Map<String, Object>> getMapList(String name)
+    {
+        return (List<Map<String, Object>>) get(name);
+    }
+    
+    public Map<String, String> getStringMap(String name)
+    {
+        final Map<String, String> map = new HashMap<String, String>();
+        
+        for (final Entry<String, Object> entry : getNode(name).entrySet())
+        {
+            if (entry.getValue() instanceof String)
+            {
+                map.put(entry.getKey(), entry.getValue().toString());
+            }
+        }
+        
+        return map;
+    }
+    
+    public Map<String, Object> getMap(String name)
+    {
+        return getNode(name);
+    }
+    
+    @SuppressWarnings("unchecked")
+    public Map<String, Object> getNode(String name)
+    {
+        return (Map<String, Object>) get(name);
+    }
+    
+    public boolean getBoolean(String name)
+    {
+        return getBoolean(name, false);
+    }
+    
+    public boolean getBoolean(String name, boolean def)
+    {
+        return (Boolean) (get(name) == null ? def : get(name));
+    }
+    
+    public int getInteger(String name)
+    {
+        return (Integer) get(name);
+    }
+    
+    @SuppressWarnings("unchecked")
+    public Object get(String nodeName)
+    {
+        Object result = null;
+        
+        if (config.containsKey(nodeName))
+        {
+            result = config.get(nodeName);
+        }
+        else
+        {
+            if (nodeName.contains("."))
+            {
+                final String[] nodes = nodeName.split("\\.");
+                Object currentNode = null;
+                
+                for (final String node : nodes)
+                {
+                    if (currentNode == null)
+                    {
+                        currentNode = config.get(node);
+                    }
+                    else
+                    {
+                        if (currentNode instanceof Map)
+                        {
+                            currentNode = ((Map<String, Object>) currentNode)
+                                    .get(node);
+                        }
+                        else
+                        {
+                            break;
+                        }
+                    }
+                }
+                
+                result = currentNode;
+            }
+            else
+            {
+                result = config.get(nodeName);
+            }
+        }
+        
+        return result;
+    }
+    
+    public Map<String, Object> getConfig()
+    {
+        return config;
+    }
+    
+    public boolean has(String node)
+    {
+        return get(node) != null;
+    }
+    
+    public void clear()
+    {
+        config.clear();
+    }
+    
+    public void save(File file) throws Exception
+    {
+        final OutputStream out = new FileOutputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at line 299. The resource is closed at line 300. There are other execution paths that do not contain closure statements, for example, when OutputStream.write throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).